src/codegen/spirv.zig

```diff
@@ -49,31 +55,326 @@ pub const SPIRVModule = struct {
+            /// which the relevant capability is not enabled are not emulated.
+            float,
+
+            /// An integer of a 'strange' size (which' bit size is not the same as its backing type. **Note**: this
+            /// may **also** include power-of-2 integers for which the relevant capability is not enabled), but still
+            /// within the limits of the largest natively supported integer type.
+            strange_integer,
+
+            /// An integer with more bits than the largest natively supported integer type.
+            composite_integer,
+        };
+
+        /// The number of bits in the inner type.
+        /// Note: this is the actual number of bits of the type, not the size of the backing integer.
+        bits: u16,
+
+        /// Whether the type is a vector.
+        is_vector: bool,
+
+        /// Whether the inner type is signed. Only relevant for integers.
+        signedness: std.builtin.Signedness,
+
+        /// A classification of the inner type. These scenarios
+        /// will all have to be handled slightly different.
+        class: Class,
+    };
+
+    fn fail(self: *DeclGen, src: LazySrcLoc, comptime format: []const u8, args: anytype) Error {
+        @setCold(true);
+        const src_loc = src.toSrcLocWithDecl(self.decl);
+        self.error_msg = try Module.ErrorMsg.create(self.module.gpa, src_loc, format, args);
+        return error.AnalysisFail;
+    }
+
+    fn resolve(self: *DeclGen, inst: *Inst) !u32 {
+        if (inst.value()) |val| {
+            return self.genConstant(inst.ty, val);
+        }
+
+        return self.values.get(inst).?; // Instruction does not dominate all uses!
+    }
+
+    /// SPIR-V requires enabling specific integer sizes through capabilities, and so if they are not enabled, we need
+    /// to emulate them in other instructions/types. This function returns, given an integer bit width (signed or unsigned, sign
+    /// included), the width of the underlying type which represents it, given the enabled features for the current target.
+    /// If the result is `null`, the largest type the target platform supports natively is not able to perform computations using
+    /// that size. In this case, multiple elements of the largest type should be used.
+    /// The backing type will be chosen as the smallest supported integer larger or equal to it in number of bits.
+    /// The result is valid to be used with OpTypeInt.
+    /// TODO: The extension SPV_INTEL_arbitrary_precision_integers allows any integer size (at least up to 32 bits).
+    /// TODO: This probably needs an ABI-version as well (especially in combination with SPV_INTEL_arbitrary_precision_integers).
+    /// TODO: Should the result of this function be cached?
+    fn backingIntBits(self: *DeclGen, bits: u16) ?u16 {
+        const target = self.module.getTarget();
+
+        // TODO: Figure out what to do with u0/i0.
```

Comment:
`hasCodeGenBits()`for u0/i0 will be `false` and so the SPIR-V backend will not be asked to lower a 0 bit integer. The assertion is correct.